File: Backtracking/nQueens.py
'''
0051 N-Queens
Must-have, so classic
# Const space? As n = 1, 2, ..., 9
# Just use it?
# Row is okay, we just loop through it
# Collections to record "placed" on column
# main diagonal: (row, col) row - col --> unique hash key
# alter diagonal: (row, col) row + col --> unique hash key
'''
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        def generateBoard():
            '''
            Out:
                board, a solution to the problem
            '''
            board = list()
            for i in range(n):
                # Here, row refers to each row of the board
                # queens is the column info
                rows[queens[i]] = 'Q'
                board.append("".join(rows))
                rows[queens[i]] = '.'
            return board
        
        def backtrack(row):
            if row == n:
                board = generateBoard()
                solutions.append(board)
            else:
                # Loop through possible column position
                # At current row
                for j in range(n):
                    # j -> columnSet
                    # row - j -> main diagonal
                    # row + j -> alt diagonal
                    if j in columnSet or row - j in diag1 or row + j in diag2:
                        continue
                    # Possible location
                    queens[row] = j
                    columnSet.add(j)
                    diag1.add(row - j)
                    diag2.add(row + j)
                    backtrack(row + 1)
                    # backtrack
                    diag1.remove(row - j)
                    diag2.remove(row + j)
                    columnSet.remove(j)
        solutions = list()
        columnSet = set()
        diag1 = set()
        diag2 = set()
        queens = [-1] * n
        rows = ['.'] * n
        backtrack(0)
        return solutions

File: Backtracking/N-queens.cpp
/**
 * 0051 N-Queens
 * Cpp version.
 */

class Solution 
{
public:
    vector<vector<string>> solveNQueens(int n) 
    {
        // Prepare the sets we need. Auto just for simplicity. 
        auto solutions = vector<vector<string>>(); // Result
        auto queens = vector<int> (n, -1); // Column position of each queen
        auto columns = unordered_set<int>();
        auto diagonals = unordered_set<int>();
        auto offDiagonals = unordered_set<int>();
        backtrack(solutions, queens, n, 0, columns, diagonals, offDiagonals);
        return solutions;
    }
    
    void backtrack(vector<vector<string>>& solutions, vector<int>& queens, int n, int row, unordered_set<int> columns, unordered_set<int> diagonals, unordered_set<int> offDiagonals) 
    {
        // Termination point
        if (row == n) 
        {
            vector<string> board = generateBoard(queens, n);
            solutions.push_back(board);
            return;
        }
        // Traverse through columns
        for (int i = 0; i < n; i++) 
        {
            // Column check
            if (columns.find(i) != columns.end()) 
            {
                continue;
            }
            // Diagonal check
            if (diagonals.find(row - i) != diagonals.end()) 
            {
                continue;
            }
            if (offDiagonals.find(row + i) != offDiagonals.end()) 
            {
                continue;
            }
            queens[row] = i; // Try this position
            columns.insert(i);
            diagonals.insert(row - i);
            offDiagonals.insert(row + i);
            backtrack(solutions, queens, n, row + 1, columns, diagonals, offDiagonals);
            queens[row] = -1; // Backtrack
            columns.erase(i);
            diagonals.erase(row - i);
            offDiagonals.erase(row + i);
        }
    }
    
    vector<string> generateBoard(vector<int>& queens, int n) 
    {
        auto board = vector<string>();
        for (int i = 0; i < n; i++) 
        {
            string row = string(n, '.');
            row[queens[i]] = 'Q';
            board.push_back(row);
        }
        return board;
    }
};